lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java

```diff
@@ -279,7 +297,130 @@ public void collect(int doc) throws IOException {
+   * globally. Post that, a global calibration step is performed
+   */
+   public static class EarlyTerminatingFieldCollector extends TopFieldCollector {
+
+    final Sort sort;
+    final FieldValueHitQueue<Entry> queue;
+    final EarlyTerminatingFieldCollectorManager earlyTerminatingFieldCollectorManager;
+    private final AtomicInteger globalNumberOfHits;
+    private boolean addedSelfToGlobalQueue;
+
+    //TODO: Refactor this to make an interface only for field collector uses
+    public EarlyTerminatingFieldCollector(Sort sort, FieldValueHitQueue<Entry> queue, int numHits, int totalHitsThreshold,
+                                          EarlyTerminatingFieldCollectorManager collectorManager, AtomicInteger globalNumberOfHits) {
+      super(queue, numHits, totalHitsThreshold, sort.needsScores());
+      this.sort = sort;
+      this.queue = queue;
+      this.earlyTerminatingFieldCollectorManager = collectorManager;
+      this.globalNumberOfHits = globalNumberOfHits;
+    }
+
+    @Override
+    public LeafCollector getLeafCollector(LeafReaderContext context) throws IOException {
+      docBase = context.docBase;
+
+      final LeafFieldComparator[] comparators = queue.getComparators(context);
+      final int[] reverseMul = queue.getReverseMul();
+      final Sort indexSort = context.reader().getMetaData().getSort();
+      final boolean canEarlyTerminate = canEarlyTerminate(sort, indexSort);
+
+      return new EarlyTerminatingMultiComparatorLeafCollector(comparators, reverseMul, this) {
+
+        boolean collectedAllCompetitiveHits = false;
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {
+          super.setScorer(scorer);
+          updateMinCompetitiveScore(scorer);
+        }
+
+        @Override
+        public void collect(int doc) throws IOException {
+
+          if (globalNumberOfHits.incrementAndGet() > numHits) {
+            if (addedSelfToGlobalQueue == false) {
+              Entry returnedEntry = earlyTerminatingFieldCollectorManager.addCollectorToGlobalQueue(earlyTerminatingFieldCollector, docBase);
+
+              if (returnedEntry != null) {
+                filterCompetitiveHit(returnedEntry.doc, false, returnedEntry.values);
+
+                if (queue.size() > 0) {
+                  Entry entry = queue.pop();
+
+                  while (entry != null) {
+                    filterCompetitiveHit(entry.doc, false, entry.values);
+                    entry = queue.pop();
+                  }
+                }
+              }
+              addedSelfToGlobalQueue = true;
+            }
+
+            filterCompetitiveHit(doc, true, comparator.leafValue(doc));
+          } else {
+            // Startup transient: queue hasn't gathered numHits yet
+            int slot = totalHits;
+            ++totalHits;
+
+            comparator.copy(slot, doc);
+            add(slot, doc, comparator.leafValue(doc));
```

Comment:
The fact that we need `leafValue` on every insert will make this threaded implementation more costly ... but that is a good tradeoff for use case that benefit from multi-threaded search.